### PR TITLE
Reconcile bindings against set of nodes reported by target cluster

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

This PR aims to complete the work of #271. 

- [x] enhance EndpointRegistry to be capable of reconciling a virtual cluster, binding/unbinding additional ports as necessary
- [x] add `BrokerPerPort` ClusterEndpointConfigProvider, retiring the `Static` implementation
- [x] remove the notion of ÇECP#getNumberOfBrokerEndpointsToPrebind 
- [x] BrokerAddressFilter becomes an internal concern and takes responsibility for initiating the endpoint reconciliation.

_Please describe your pull request_

### Additional Context

_Why are you making this pull request?_

### CODE REVIEW

**Comment by k-wall** [comment]
@robobario @SamBarker the first part of the work is complete, so I am marking as ready for review to get some early feedback.  I'll continue with the 'handler' work next.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java [hunk 1]
```diff
@@ -167,11 +174,27 @@ private void addHandlers(SocketChannel ch, VirtualClusterBinding binding) {
...
+                    }
+                    if (b.nodeId() >= 0) {
+                        currentNodeIds.add(b.nodeId());
+                    }
+                    else {
+                        // TODO - KW not sure if we actually need this guard
+                        LOGGER.warn("Got a unexpected node id {} - ignored it", b.nodeId());
+                    }
+                });
+                endpointReconciler.reconcile(virtualCluster, currentNodeIds).whenComplete((unused, t) -> {
```
Comment (k-wall) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java:192]: I'll be refactoring this part next, and adding some unit tests.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java [hunk 2]
```diff
@@ -238,6 +249,100 @@ public CompletionStage<Void> deregisterVirtualCluster(VirtualCluster virtualClus
...
+        else {
+            // set of nodeIds differ
+            return rec.getValue().thenCompose(u -> {
+                Map.Entry<Set<Integer>, CompletionStage<Void>> updated;
+                var cand = createReconcileEntry(nodeIds, new CompletableFuture<>());
+                if (vcr.reconciliationEntry().compareAndSet(rec, cand)) {
+                    // reconcile - work out which bindings are to be registered and which are to be removed.
+                    var bindingAddress = virtualCluster.getBindAddress().orElse(null);
+
+                    var toCreate = nodeIds.stream().map(i -> new VirtualClusterBrokerBinding(virtualCluster, i))
```
Comment (robobario) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java:291]: Would this be improved by planning it out first? Collect up the set of existing VCBB, classify them into `toRemove`, `toCreate`,`toLeave` then kick off the futures?

**Review by robobario — COMMENTED** [review]
Pulling the YAML unit-testing code into test-tools is great thanks @k-wall

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java [hunk 3]
```diff
@@ -238,6 +249,100 @@ public CompletionStage<Void> deregisterVirtualCluster(VirtualCluster virtualClus
...
+                                                .map(vcbb -> deregisterBinding(virtualCluster, vcbb::equals)));
+                                    }))));
+
+                    // chain any binding registrations and organise for the reconciliations entry to complete
+                    var unused = deregs.thenCompose(u1 -> allOfStage(toCreate.stream()
+                            .map(vcbb -> {
+                                var brokerAddress = virtualCluster.getBrokerAddress(vcbb.nodeId());
+                                var endpoint = new Endpoint(bindingAddress, brokerAddress.port(),
+                                        virtualCluster.isUseTls());
+                                return registerBinding(endpoint, brokerAddress.host(), vcbb);
```
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java:321]: I am quite possibly missing something. However I feel like a binding should know the endpoint. Otherwise what is it binding to?

**Review by SamBarker — COMMENTED** [review]
I'm still trying to follow the reconciliation flow, will pick that up again tomorrow however publishing the thoughts I have now to avoid loosing them.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                                                          @JsonProperty(required = false) Integer brokerStartPort,
+                                                          @JsonProperty(required = false, defaultValue = "3") Integer numberOfBrokerPorts) {
+            Objects.requireNonNull(bootstrapAddress, "bootstrapAddress cannot be null");
+
+            this.bootstrapAddress = bootstrapAddress;
+            this.brokerAddressPattern = brokerAddressPattern != null ? brokerAddressPattern : (bootstrapAddress.host() + ":" + LITERAL_PORT_NUMBER);
+            this.brokerStartPort = brokerStartPort != null ? brokerStartPort : (bootstrapAddress.port() + 1);
+            this.numberOfBrokerPorts = numberOfBrokerPorts != null ? numberOfBrokerPorts : 3;
+
+            if (this.brokerStartPort < 1) {
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:105]: yeah - I'm been having these ideas too.  I started to poke around JSR 380 with Jackson yesterday, but I figure this is a separate piece of work that should be applied to the whole object model.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * Creates the configuration for this provider.
+     */
+    public static class PortPerBrokerClusterEndpointProviderConfig extends BaseConfig {
+        private final HostPort bootstrapAddress;
+        private final String brokerAddressPattern;
+        private final int brokerStartPort;
+        private final int numberOfBrokerPorts;
+
+        public PortPerBrokerClusterEndpointProviderConfig(@JsonProperty(required = true) HostPort bootstrapAddress,
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:94]: Could you elaborate?

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java [hunk 6]
```diff
@@ -98,16 +99,25 @@ private SniRoutingKey(String sniHostname) {
 
     protected static final AttributeKey<Map<RoutingKey, VirtualClusterBinding>> CHANNEL_BINDINGS = AttributeKey.newInstance("channelBindings");
 
-    private record VirtualClusterRecord(CompletionStage<Endpoint> registrationStage, AtomicReference<CompletionStage<Void>> deregistrationStage) {
+    private record VirtualClusterRecord(CompletionStage<Endpoint> registrationStage, AtomicReference<Map.Entry<Set<Integer>, CompletionStage<Void>>> reconciliationEntry, AtomicReference<CompletionStage<Void>> deregistrationStage) {
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java:102]: I think you are referring to the use of the Entry, and that this ought to be an object in its own right.  If so, yes, that would probably be cleaner.

The code relies on the CAS semantics for thread exclusivity, so the surrounding AtomicReference would still be required.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Override
+    public HostPort getClusterBootstrapAddress() {
+        return this.bootstrapAddress;
+    }
+
+    @Override
+    public HostPort getBrokerAddress(int nodeId) throws IllegalArgumentException {
+        int port = brokerStartPort + nodeId;
+        if (port >= brokerStartPort + numberOfBrokerPorts) {
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java]: Done.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:103]: The `bootstrap.servers` property usually contains multiple hosts so I was kinda thinking we would have that here as well.

**Comment by robobario** [comment]
A late question on the whole solution, what do we get from dynamically binding/unbinding the ports as brokers come and go in the metadata? If there are 3 nodes and we allocate a port range for 5 nodes to allow for growth, could we bind all 5 ports up front? 

Wondering if this might remove some race conditions like kroxylicious instance X being aware of broker-4 before kroxylicious instance Y, if they are bound up front we could connect to kroxylicious instance Y on the broker-4 port before it's aware of broker-4 via metadata.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            var matcher = ANCHORED_PORT_NUMBER_TOKEN_RE.matcher(this.brokerAddressPattern);
+            if (!matcher.find()) {
+                throw new IllegalArgumentException("brokerAddressPattern must contain exactly one replacement pattern " + LITERAL_PORT_NUMBER + ". Found none.");
+            }
+            var stripped = matcher.replaceFirst("");
+            matcher = ANCHORED_PORT_NUMBER_TOKEN_RE.matcher(stripped);
+            if (matcher.find()) {
+                throw new IllegalArgumentException(
+                        "brokerAddressPattern must contain exactly one replacement pattern " + LITERAL_PORT_NUMBER + ". Found too many.");
+            }
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:134]: The overhead of introducing a utility class seemed to outweigh benefits of removing this small duplication.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Override
+    public HostPort getBrokerAddress(int nodeId) throws IllegalArgumentException {
+        int port = brokerStartPort + nodeId;
+        if (port >= brokerStartPort + numberOfBrokerPorts) {
+            throw new IllegalArgumentException(
+                    "Cannot generate broker address for node id %d port range (%d - %d)".formatted(nodeId, brokerStartPort, numberOfBrokerPorts));
+        }
+        return HostPort.parse(brokerAddressPattern.replace(LITERAL_PORT_NUMBER, Integer.toString(port)).replace(LITERAL_NODE_ID, Integer.toString(nodeId)));
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:86]: Please take another look now.   `BrokerAddressFilter` is now using only `EndpointReconciler` to trigger the reconciliation with the upstream nodeids/upstream addresses.   It still needs `VirtualCluster#getBrokerAddress` to get the downstream mapping.

It is suggesting to me that `EndpointRegistry` should have responsible for providing the downstream addresses too - perhaps a method `EndpointRegistry#getDownstreamBrokerMap(virtualCluster)` -> `Map<Integer, HostPort>`.  Internally the registry could cache the downstreamTarget in the binding.   I think that would address your comment about caching.   WDYT?  (I'm inclined to experiment with this on subsequent PR btw)

It also put us into a place where an `EndpointRegistry` implementation could start supporting dynamic ports - the CECP would be define a downstream address as port 0, the endpoint registry would cache the resulting dynamically assigned port from the network stack.  Dynamic port support would be good for kroxy's own integration test and kroxy use-case were it is embedded for short life applications.  This is not a now thing, but I wanted to voice the idea.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    private final int brokerStartPort;
+    private final int numberOfBrokerPorts;
+    private final Set<Integer> exclusivePorts;
+
+    /**
+     * Creates the provider.
+     *
+     * @param config configuration
+     */
+    public PortPerBrokerClusterEndpointConfigProvider(PortPerBrokerClusterEndpointProviderConfig config) {
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:55]: In Kafka, clients only ever talk to nodes that are in the broker role, so I think 'broker' is correct in this case..

I'm using the term `Endpoint` to mean a network listener.  That comprises a `bindingAddress` and `portNumber`. I've also kludged in `tls` (it was convenient) .  Endpoints are used to actually make the bind to the network. Also when a connection is made, we use the get the [localAddress from the socket](https://github.com/kroxylicious/kroxylicious/blob/7c724e4716d0740b85220a99522c6e5f4bf9af8f/kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java#L64)  use that to create an Endpoint object.  That's the key to the EndpointRegistry.

The `HostPort` is really the network address from client perspective.  There are three use-cases 1) we give the client the host/port in the MetadataResponse etc, 2) we use the _host_ part for SNI resolution 3) we use the _port_ part to create the endpoint.



  We need HostPort for two reasons - firstly that's the tuple that needs to go to the kafka client in the metadata response - and second we need to know the 

In the CECP the HostPort

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java (hunk 3 above)
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java]: Originally, my idea was that the binding would provide the tuple (virtualcluster, brokerid) and we'd use the brokerid to lookup the upstream or downstream address.  In this PR, I've now included the upstreamTarget address in the binding.  I've left comments above above whether EndpointRegistry should take responsibility for downstream addresses too.
 Please see the comments on the subject above.

**Comment by k-wall** [comment]
> A late question on the whole solution, what do we get from dynamically binding/unbinding the ports as brokers come and go in the metadata? If there are 3 nodes and we allocate a port range for 5 nodes to allow for growth, could we bind all 5 ports up front?
> 
> Wondering if this might remove some race conditions like kroxylicious instance X being aware of broker-4 before kroxylicious instance Y, if they are bound up front we could connect to kroxylicious instance Y on the broker-4 port before it's aware of broker-4 via metadata.

Doesn't it just change the race condition, rather than eliminate it?  Kroxy would still need to discover the upstream address mapping and in the case where kroxy is deployed in a pool, that would occur over time.

**Comment by robobario** [comment]
> Doesn't it just change the race condition, rather than eliminate it? Kroxy would still need to discover the upstream address mapping and in the case where kroxy is deployed in a pool, that would occur over time.

D'oh, I see, I'd lost track of the fact we are using the metadata to discover the upstream address. Maybe a link to #271 in the PR description would help future readers.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:86]: > It is suggesting to me that EndpointRegistry should have responsible for providing the downstream addresses too - perhaps a method EndpointRegistry#getDownstreamBrokerMap(virtualCluster) -> Map<Integer, HostPort>. Internally the registry could cache the downstreamTarget in the binding. I think that would address your comment about caching. WDYT? (I'm inclined to experiment with this on subsequent PR btw)

I think you're moving in a similar direction to me (hence my comment about the wrong level). I'm happy to proceed with this PR without sorting a cache but I do think its important to the overall performance profile.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:55]: >In Kafka, clients only ever talk to nodes that are in the broker role, so I think 'broker' is correct in this case..
We have loosely discussed use cases where Kroxylicious is injected between brokers as well as clients, though I don't know if any of those thoughts are really realistic.

I'm ok with keeping both `Endpoint` and `HostPort` classes (I'm not sure I get the distinction you're drawing). The class in question is called `PortPerBrokerClusterEndpointConfigProvider` but no where in the API surface do we see `Endpoint` but we do see `HostPort` which is actually my larger point. If it's providing configuration for Endpoints as per the name then I think it should be returning Endpoints, if it's providing something else then I feel the name should change.

[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:55]: I am getting your point and a do believe naming matters.  Let's keep talking to see if we can make things better.

The `ClusterEndpointConfigProvider` is providing all the networky configuration a virtual cluster needs.  There are the three aspects to this:

1.  Provides information to let kroxy bind a listening sockets
2. Provides information to let kafka clients know how to dial Kroxy to talk to the brokers of a virtual cluster.
3. For the SNI case, let kroxy route connections to their correct destinations.

So at an API level:

* ClusterEndpointConfigProvider#getBindAddress required by **1**.
* ClusterEndpointConfigProvider#getClusterBootstrapAddress - the port portion is required by **1**.  the host portion is required by **3**. 
* ClusterEndpointConfigProvider#getBrokerAddress  - the port portion is required by **1**.  the host portion is required by **3**, the host/port portion is required by **2**.

I don't really see a way to tease apart this information into separate domain objects, so I think I think it belongs together in one.

How about `ClusterNetworkConfigProvider`?

[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:55]: `ClusterNetworkConfigProvider` is certainly a lot better! 

My only passing thought: is `Network` too generic? Maybe `NetworkAdress`? 

I'm happy to go with wich ever you think is better.

**Comment by k-wall** [comment]
@SamBarker @robobario could you re-review please?  I'm hoping that this is approaching mergeable.

My thinking is that we live this things in this shape for a week or so and give ourselves chance to reflect on the next step (should we split endpointregistry?, should endpointregistry be the cache for upstream addresses etc).   I'm also very keen that someone apart from me should take the next step.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusterendpointprovider/PortPerBrokerClusterEndpointConfigProvider.java:103]: I think I was assuming `bootstrapAddress` was the dial address for the upstream cluster, hence why a single value seems odd. 

How about: 
```suggestion
        public PortPerBrokerClusterEndpointProviderConfig(@JsonProperty(required = true) HostPort bootstrapBindAddress,
```

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/filter/BrokerAddressFilter.java [hunk 11]
```diff
@@ -59,12 +59,18 @@ public void onDescribeClusterResponse(ResponseHeaderData header, DescribeCluster
...
     public void onFindCoordinatorResponse(ResponseHeaderData header, FindCoordinatorResponseData data, KrpcFilterContext context) {
+        // Version 4+
         for (Coordinator coordinator : data.coordinators()) {
             // If the coordinator is not yet available, the server returns a nodeId of -1.
             if (coordinator.nodeId() >= 0) {
                 apply(context, coordinator, Coordinator::nodeId, Coordinator::host, Coordinator::port, Coordinator::setHost, Coordinator::setPort);
             }
         }
+        // Version 3
+        if (data.nodeId() >= 0 && data.host() != null && !data.host().isEmpty() && data.port() > 0) {
```
[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/filter/BrokerAddressFilter.java:82]: The bigger thought about API version handling is covered by https://github.com/kroxylicious/kroxylicious/pull/338. Lets leave this as is until 338 lands and deal with the migration then.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java [hunk 12]
```diff
@@ -209,6 +217,14 @@ public void forwardResponse(ApiMessage response) {
         }
     }
 
+    @Override
+    public void abortConnection(Throwable t) {
```
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java:221]: I think the usage of this makes sense, but I'm not sure we want to expose `abortConnection` to all filter authors as part of the public API.

I think there is a wider conversation to be had around how we handle failure in filters and how filters may want to signal that a connection is in a bad state. 

Could we do something hacky and tactical for now such as casting to `DefaultFilterContext` at the call site for access to abort connection? Or are we happy to add this to the public API and remove later if we decide on an exception based flow?

**Comment by robobario** [comment]
I think the `EndpointRegistry` and `EndpointReconciler` interfaces work and we can refactor the implementations later. 

One thing I'm wondering is if it would improve anything by instead making a queue of commands within EndpointRegistry. So all the public methods would drop a command in a queue like `RegisterCluster(VirtualCluster cluster)`, `ReconcileCluster(VirtualCluster virtualCluster, Map<Integer, HostPort> upstreamNodes)`. A single thread would pull them off the queue and apply the logic. Naively we could block on the bind/unbind commands. We could then stop tracking which other operations are in progress and have objects like Channel directly in the maps.

I think with this setup we would have easier code to understand but could be hurt more by any slowness, since reconciliation of one cluster could block reconciliation of another.

**Review by SamBarker — APPROVED** [review]
I've found the reconciliation flow hard to follow and comprehend however I spent a bunch of time trying to tease out something which made more sense to me and all have to show for it is an appreciation of the subtle complexity in what is being done. 

For the record, I think we are missing a concept or are trying to tie distinct flows together however I can't point at anything specific and say we should do some thing else here. 

I think dealing in futures is the right thing to do for the API and I'm comfortable with that side of things. 

So I'm for merging this and then revisiting the reconciliation flow. 

My only hesitation is around exposing [abortConnection](https://github.com/kroxylicious/kroxylicious/pull/324/files#r1208671696) in the public FilterContext API.

**Review by robobario — APPROVED** [review]

**Comment by robobario** [comment]
Here's a PR with [main](https://github.com/k-wall/kroxylicious/pull/2) merged in. To do this I introduced a `kroxylicious-unit-test` module to avoid a cyclic dependency between kroxylicious and kroxylicious-test-tools. Also it showed up a possible issue with doing asynchronous work during the Metadata response which is addressed in the PR.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java (hunk 12 above)
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java]: After discussion, we switched to synchronous reconciliation to side step #351 so the immediate need for `abortConnection` goes.
